File: app/src/main/java/io/github/benoitduffez/cupsprint/app/HostNotVerifiedActivity.java
package io.github.benoitduffez.cupsprint.app;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import io.github.benoitduffez.cupsprint.R;

/**
 * Ask for host trust when it couldn't be verified
 */
public class HostNotVerifiedActivity extends Activity {
    public static final String KEY_HOST = HostNotVerifiedActivity.class.getName() + ".ErrorText";

    private static final String HOSTS_FILE = "hosts_trust";

    private String mUnverifiedHostname;

    /**
     * Check whether host is known and trusted
     *
     * @param context  Used to retrieve the saved setting
     * @param hostname Hostname to be checked
     * @return true if the hostname was explicitly trusted, false otherwise or if unknown
     */
    public static boolean isHostnameTrusted(Context context, String hostname) {
        return context.getSharedPreferences(HOSTS_FILE, MODE_PRIVATE).getBoolean(hostname, false);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.host_not_verified);
        getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);

        TextView title = (TextView) findViewById(R.id.host_not_verified_title);
        mUnverifiedHostname = getIntent().getStringExtra(KEY_HOST);
        title.setText(getString(R.string.host_not_verified_title, mUnverifiedHostname));

        findViewById(R.id.host_not_verified_trust_button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                validateTrust(true);
            }
        });
        findViewById(R.id.host_not_verified_abort_button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                validateTrust(false);
            }
        });
    }

    /**
     * Save choice and finish
     *
     * @param trusted Whether the host should be trusted or not
     */
    void validateTrust(boolean trusted) {
        SharedPreferences.Editor prefs = getSharedPreferences(HOSTS_FILE, MODE_PRIVATE).edit();
        prefs.putBoolean(mUnverifiedHostname, trusted);
        prefs.apply();
        finish();
    }
}

File: app/src/main/java/io/github/benoitduffez/cupsprint/app/HostNotVerifiedActivity.kt
package io.github.benoitduffez.cupsprint.app

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.ViewGroup
import io.github.benoitduffez.cupsprint.R
import kotlinx.android.synthetic.main.host_not_verified.*

/**
 * Ask for host trust when it couldn't be verified
 */
class HostNotVerifiedActivity : Activity() {
    private var unverifiedHostname: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.host_not_verified)
        window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)

        unverifiedHostname = intent.getStringExtra(KEY_HOST)
        host_not_verified_title.text = getString(R.string.host_not_verified_title, unverifiedHostname)

        host_not_verified_trust_button.setOnClickListener { validateTrust(true) }
        host_not_verified_abort_button.setOnClickListener { validateTrust(false) }
    }

    /**
     * Save choice and finish
     *
     * @param trusted Whether the host should be trusted or not
     */
    private fun validateTrust(trusted: Boolean) {
        val prefs = getSharedPreferences(HOSTS_FILE, Context.MODE_PRIVATE).edit()
        prefs.putBoolean(unverifiedHostname, trusted)
        prefs.apply()
        finish()
    }

    companion object {
        val KEY_HOST = "${HostNotVerifiedActivity::class.java.name}.ErrorText"

        private const val HOSTS_FILE = "hosts_trust"

        /**
         * Check whether host is known and trusted
         *
         * @param context  Used to retrieve the saved setting
         * @param hostname Hostname to be checked
         * @return true if the hostname was explicitly trusted, false otherwise or if unknown
         */
        fun isHostnameTrusted(context: Context, hostname: String): Boolean {
            return context.getSharedPreferences(HOSTS_FILE, Context.MODE_PRIVATE).getBoolean(hostname, false)
        }
    }
}
